File: ts/indicators/john-bollinger/bollinger-bands-trend/bollinger-bands-trend.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, DefaultBarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { QuoteComponent, DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { TradeComponent, DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { ExponentialMovingAverage } from '../../common/exponential-moving-average/exponential-moving-average';
import { SimpleMovingAverage } from '../../common/simple-moving-average/simple-moving-average';
import { Variance } from '../../common/variance/variance';
import { BollingerBandsTrendParams, MovingAverageType } from './params';

/** Interface for an indicator that accepts a scalar and returns a value. */
interface LineUpdater {
  update(sample: number): number;
  isPrimed(): boolean;
}

/** Holds the sub-components for one Bollinger Band calculation. */
class BBLine {
  private readonly ma: LineUpdater;
  private readonly variance_: Variance;
  private readonly upperMultiplier_: number;
  private readonly lowerMultiplier_: number;
  private primed_ = false;

  constructor(
    length: number,
    upperMultiplier: number,
    lowerMultiplier: number,
    isUnbiased: boolean,
    maType: MovingAverageType | undefined,
    firstIsAverage: boolean,
    barComponent?: BarComponent,
    quoteComponent?: QuoteComponent,
    tradeComponent?: TradeComponent,
  ) {
    this.upperMultiplier_ = upperMultiplier;
    this.lowerMultiplier_ = lowerMultiplier;

    this.variance_ = new Variance({
      length,
      unbiased: isUnbiased,
      barComponent,
      quoteComponent,
      tradeComponent,
    });

    if (maType === MovingAverageType.EMA) {
      this.ma = new ExponentialMovingAverage({ length, firstIsAverage });
    } else {
      this.ma = new SimpleMovingAverage({ length });
    }
  }

  /** Returns [lower, middle, upper, primed]. */
  update(sample: number): [number, number, number, boolean] {
    const middle = this.ma.update(sample);
    const v = this.variance_.update(sample);

    this.primed_ = this.ma.isPrimed() && this.variance_.isPrimed();

    if (isNaN(middle) || isNaN(v)) {
      return [NaN, NaN, NaN, this.primed_];
    }

    const stddev = Math.sqrt(v);
    const upper = middle + this.upperMultiplier_ * stddev;
    const lower = middle - this.lowerMultiplier_ * stddev;

    return [lower, middle, upper, this.primed_];
  }
}

/**
 * John Bollinger's Bollinger Bands Trend indicator.
 *
 * BBTrend measures the difference between the widths of fast and slow Bollinger Bands
 * relative to the fast middle band, indicating trend strength and direction.
 *
 * The indicator produces a single output:
 *
 *   bbtrend = (|fastLower - slowLower| - |fastUpper - slowUpper|) / fastMiddle
 *
 * Reference:
 *
 * Bollinger, John (2002). Bollinger on Bollinger Bands. McGraw-Hill.
 */
export class BollingerBandsTrend implements Indicator {

  private readonly fastBB: BBLine;
  private readonly slowBB: BBLine;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  private value_ = NaN;
  private primed_ = false;

  private readonly mnemonic_: string;
  private readonly description_: string;

  constructor(params: BollingerBandsTrendParams) {
    const fastLength = Math.floor(params.fastLength);
    const slowLength = Math.floor(params.slowLength);

    if (fastLength < 2) {
      throw new Error('fast length should be greater than 1');
    }

    if (slowLength < 2) {
      throw new Error('slow length should be greater than 1');
    }

    if (slowLength <= fastLength) {
      throw new Error('slow length should be greater than fast length');
    }

    const upperMultiplier = params.upperMultiplier ?? 2.0;
    const lowerMultiplier = params.lowerMultiplier ?? 2.0;
    const isUnbiased = params.isUnbiased ?? true;
    const firstIsAverage = params.firstIsAverage ?? false;

    const bc = params.barComponent ?? DefaultBarComponent;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    this.fastBB = new BBLine(
      fastLength, upperMultiplier, lowerMultiplier, isUnbiased,
      params.movingAverageType, firstIsAverage,
      params.barComponent, params.quoteComponent, params.tradeComponent,
    );

    this.slowBB = new BBLine(
      slowLength, upperMultiplier, lowerMultiplier, isUnbiased,
      params.movingAverageType, firstIsAverage,
      params.barComponent, params.quoteComponent, params.tradeComponent,
    );

    const cm = componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent);
    this.mnemonic_ = `bbtrend(${fastLength},${slowLength},${Math.floor(upperMultiplier)},${Math.floor(lowerMultiplier)}${cm})`;
    this.description_ = 'Bollinger Bands Trend ' + this.mnemonic_;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean {
    return this.primed_;
  }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.BollingerBandsTrend,
      this.mnemonic_,
      this.description_,
      [
        { mnemonic: this.mnemonic_, description: this.description_ },
      ],
    );
  }

  /** Updates the indicator given the next sample value and returns the BBTrend value. */
  public update(sample: number): number {
    if (isNaN(sample)) {
      return NaN;
    }

    const [fastLower, fastMiddle, fastUpper, fastPrimed] = this.fastBB.update(sample);
    const [slowLower, , slowUpper, slowPrimed] = this.slowBB.update(sample);

    this.primed_ = fastPrimed && slowPrimed;

    if (!this.primed_ || isNaN(fastMiddle) || isNaN(fastLower) || isNaN(slowLower)) {
      this.value_ = NaN;
      return NaN;
    }

    const epsilon = 1e-10;
    const lowerDiff = Math.abs(fastLower - slowLower);
    const upperDiff = Math.abs(fastUpper - slowUpper);

    if (Math.abs(fastMiddle) < epsilon) {
      this.value_ = 0;
      return 0;
    }

    const result = (lowerDiff - upperDiff) / fastMiddle;
    this.value_ = result;

    return result;
  }

  /** Updates the indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    const v = this.update(sample.value);
    const s = new Scalar();
    s.time = sample.time;
    s.value = v;
    return [s];
  }

  /** Updates the indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    const v = this.barComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    const v = this.quoteComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }

  /** Updates the indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    const v = this.tradeComponentFunc(sample);
    const scalar = new Scalar();
    scalar.time = sample.time;
    scalar.value = v;
    return this.updateScalar(scalar);
  }
}

File: go/indicators/johnbollinger/bollingerbandstrend/bollingerbandstrend.go
package bollingerbandstrend

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/common/exponentialmovingaverage"
	"zpano/indicators/common/simplemovingaverage"
	"zpano/indicators/common/variance"
	"zpano/indicators/core"
)

// lineUpdater is an interface for indicators that accept a single scalar and return a value.
type lineUpdater interface {
	Update(float64) float64
	IsPrimed() bool
}

// bbLine holds the sub-components for one Bollinger Band calculation.
type bbLine struct {
	ma              lineUpdater
	variance        *variance.Variance
	upperMultiplier float64
	lowerMultiplier float64
}

// update feeds a sample and returns lower, middle, upper, primed.
//
//nolint:nonamedreturns
func (s *bbLine) update(sample float64) (lower, middle, upper float64, primed bool) {
	nan := math.NaN()

	middle = s.ma.Update(sample)
	v := s.variance.Update(sample)

	primed = s.ma.IsPrimed() && s.variance.IsPrimed()

	if math.IsNaN(middle) || math.IsNaN(v) {
		return nan, nan, nan, primed
	}

	stddev := math.Sqrt(v)
	upper = middle + s.upperMultiplier*stddev
	lower = middle - s.lowerMultiplier*stddev

	return lower, middle, upper, primed
}

// BollingerBandsTrend is John Bollinger's Bollinger Bands Trend indicator.
//
// BBTrend measures the difference between the widths of fast and slow Bollinger Bands
// relative to the fast middle band, indicating trend strength and direction.
//
// The indicator produces a single output:
//
//	bbtrend = (|fastLower - slowLower| - |fastUpper - slowUpper|) / fastMiddle
//
// Reference:
//
// Bollinger, John (2002). Bollinger on Bollinger Bands. McGraw-Hill.
type BollingerBandsTrend struct {
	mu sync.RWMutex

	fastBB *bbLine
	slowBB *bbLine

	value  float64
	primed bool

	barFunc   entities.BarFunc
	quoteFunc entities.QuoteFunc
	tradeFunc entities.TradeFunc

	mnemonic string
}

// NewBollingerBandsTrend returns an instance of the indicator created using supplied parameters.
//
//nolint:funlen,cyclop
func NewBollingerBandsTrend(p *BollingerBandsTrendParams) (*BollingerBandsTrend, error) {
	const (
		invalid           = "invalid bollinger bands trend parameters"
		fmts              = "%s: %s"
		fmtw              = "%s: %w"
		minLength         = 2
		defaultFastLength = 20
		defaultSlowLength = 50
		defaultMultiplier = 2.0
	)

	fastLength := p.FastLength
	if fastLength == 0 {
		fastLength = defaultFastLength
	}

	slowLength := p.SlowLength
	if slowLength == 0 {
		slowLength = defaultSlowLength
	}

	if fastLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "fast length should be greater than 1")
	}

	if slowLength < minLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than 1")
	}

	if slowLength <= fastLength {
		return nil, fmt.Errorf(fmts, invalid, "slow length should be greater than fast length")
	}

	upperMultiplier := p.UpperMultiplier
	if upperMultiplier == 0 {
		upperMultiplier = defaultMultiplier
	}

	lowerMultiplier := p.LowerMultiplier
	if lowerMultiplier == 0 {
		lowerMultiplier = defaultMultiplier
	}

	isUnbiased := true
	if p.IsUnbiased != nil {
		isUnbiased = *p.IsUnbiased
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	// Create fast and slow BB sub-components.
	fastBB, err := newBBLine(fastLength, upperMultiplier, lowerMultiplier, isUnbiased, p.MovingAverageType, p.FirstIsAverage, bc, qc, tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	slowBB, err := newBBLine(slowLength, upperMultiplier, lowerMultiplier, isUnbiased, p.MovingAverageType, p.FirstIsAverage, bc, qc, tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf("bbtrend(%d,%d,%.0f,%.0f%s)", fastLength, slowLength, upperMultiplier, lowerMultiplier,
		core.ComponentTripleMnemonic(bc, qc, tc))

	return &BollingerBandsTrend{
		fastBB:    fastBB,
		slowBB:    slowBB,
		value:     math.NaN(),
		barFunc:   barFunc,
		quoteFunc: quoteFunc,
		tradeFunc: tradeFunc,
		mnemonic:  mnemonic,
	}, nil
}

//nolint:cyclop
func newBBLine(
	length int, upperMultiplier, lowerMultiplier float64, isUnbiased bool,
	maType MovingAverageType, firstIsAverage bool,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*bbLine, error) {
	v, err := variance.NewVariance(&variance.VarianceParams{
		Length:         length,
		IsUnbiased:     isUnbiased,
		BarComponent:   bc,
		QuoteComponent: qc,
		TradeComponent: tc,
	})
	if err != nil {
		return nil, err
	}

	var ma lineUpdater

	switch maType {
	case EMA:
		ema, e := exponentialmovingaverage.NewExponentialMovingAverageLength(
			&exponentialmovingaverage.ExponentialMovingAverageLengthParams{
				Length:         length,
				FirstIsAverage: firstIsAverage,
			})
		if e != nil {
			return nil, e
		}

		ma = ema
	default:
		sma, e := simplemovingaverage.NewSimpleMovingAverage(
			&simplemovingaverage.SimpleMovingAverageParams{Length: length})
		if e != nil {
			return nil, e
		}

		ma = sma
	}

	return &bbLine{
		ma:              ma,
		variance:        v,
		upperMultiplier: upperMultiplier,
		lowerMultiplier: lowerMultiplier,
	}, nil
}

// IsPrimed indicates whether the indicator is primed.
func (s *BollingerBandsTrend) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes the output data of the indicator.
func (s *BollingerBandsTrend) Metadata() core.Metadata {
	desc := "Bollinger Bands Trend " + s.mnemonic

	return core.BuildMetadata(
		core.BollingerBandsTrend,
		s.mnemonic,
		desc,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: desc},
		},
	)
}

// Update updates the indicator given the next sample value and returns the BBTrend value.
func (s *BollingerBandsTrend) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return math.NaN()
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	fastLower, fastMiddle, fastUpper, fastPrimed := s.fastBB.update(sample)
	slowLower, _, slowUpper, slowPrimed := s.slowBB.update(sample)

	s.primed = fastPrimed && slowPrimed

	if !s.primed || math.IsNaN(fastMiddle) || math.IsNaN(fastLower) || math.IsNaN(slowLower) {
		s.value = math.NaN()
		return math.NaN()
	}

	const epsilon = 1e-10

	lowerDiff := math.Abs(fastLower - slowLower)
	upperDiff := math.Abs(fastUpper - slowUpper)

	if math.Abs(fastMiddle) < epsilon {
		s.value = 0
		return 0
	}

	result := (lowerDiff - upperDiff) / fastMiddle
	s.value = result

	return result
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *BollingerBandsTrend) UpdateScalar(sample *entities.Scalar) core.Output {
	v := s.Update(sample.Value)

	return []any{entities.Scalar{Time: sample.Time, Value: v}}
}

// UpdateBar updates the indicator given the next bar sample.
func (s *BollingerBandsTrend) UpdateBar(sample *entities.Bar) core.Output {
	v := s.barFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *BollingerBandsTrend) UpdateQuote(sample *entities.Quote) core.Output {
	v := s.quoteFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *BollingerBandsTrend) UpdateTrade(sample *entities.Trade) core.Output {
	v := s.tradeFunc(sample)

	return s.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}
